### ISSUE TITLE
Low performance on complex recursive types.

### ISSUE THREAD

**Issue by jkb0o** [microsoft/pyright#1795]
**Describe the bug**
If I use some complex recursive type signature (like `Variant` in the example below), it takes a lot of time to analyze the code (14+ seconds for a single file, and up to 80 seconds, if I use this `Variant` in other modules). I've tested this using `pyright test.py`.

It looks like infinity to wait for `pylance` to complete the analysis. The only workaround for my case is to isolate a `Variant` to its own container type or use `Any` instead of `Variant`.

**To Reproduce**
Save example code into `test.py`, run `pyright test.py`.

**Expected behavior**
I'd like to see some constant time analysis (in less than 1 second in a perfect world).

**Code**

```python
from typing import (Dict, List, Literal, NamedTuple, NewType, Optional, Type,
                    TypeVar, Union, cast, get_args, get_origin)


class Infinity:
    pass

class NaN:
    pass

class Color(NamedTuple):
    r: float
    g: float
    b: float
    a: float

class Vector2(NamedTuple):
    x: float
    y: float
        
class Vector3(NamedTuple):
    x: float
    y: float
    z: float

class Transform2D(NamedTuple):
    x: Vector2
    y: Vector2
    origin: Vector2

class Rect2(NamedTuple):
    position: Vector2
    size: Vector2

class Plane(NamedTuple):
    normal: Vector3
    distance: float

class Quat(NamedTuple):
    x: float
    y: float
    z: float
    w: float

class AABB(NamedTuple):
    position: Vector3
    size: Vector3

class Basis(NamedTuple):
    x: Vector3
    y: Vector3
    z: Vector3

class Transform(NamedTuple):
    basis: Basis
    origin: Vector3

class NodePath(str):
    pass

class RID(int):
    pass

class Object(NamedTuple):
    object_type: str
    args: Dict[str, 'Variant']

class Resource(NamedTuple):
    resource_type: Literal["Resource", "ExtResource", "SubResource"]
    path: Union[str, int]

Byte = NewType('Byte', int)
T = TypeVar('T', Byte, int, float, str, Vector2, Vector3, Color)
class Array(List[T]):
    def __init__(self, array_type: Type[T]):
        self.array_type = array_type

Variant = Union[
    None, int, float, str,
    Type[Infinity], Type[NaN],
    Color, Vector2, Rect2, Vector3,
    Transform2D, Plane, Quat, AABB,
    Basis, Transform, NodePath, RID,
    Object, Resource, Array[Color],
    Array[Byte], Array[int], Array[float],
    Array[str], Array[Vector2], Array[Vector3],
    List['Variant'],
    Dict['Variant', 'Variant']
]

TVariant = TypeVar('TVariant', 
    None, int, float, str,
    Type[Infinity], Type[NaN],
    Color, Vector2, Rect2, Vector3,
    Transform2D, Plane, Quat, AABB,
    Basis, Transform, NodePath, RID,
    Object, Resource, Array[Color],
    Array[Byte], Array[int], Array[float],
    Array[str], Array[Vector2], Array[Vector3],
    List['Variant'],
    Dict['Variant', 'Variant']
)

def default(typ: Type[TVariant]) -> TVariant:
    if typ == int:
        return cast(TVariant, 0)
    elif typ == float:
        return cast(TVariant, 0.0)
    elif typ == str:
        return cast(TVariant, "")
    elif typ == Color:
        return cast(TVariant, Color(1, 1, 1, 1))
    elif typ == Vector2:
        return cast(TVariant, Vector2(0, 0))
    elif typ == Vector3:
        return cast(TVariant, Vector3(0, 0, 0))
    elif typ == Rect2:
        return cast(TVariant, Rect2(Vector2(0, 0), Vector2(0, 0)))
    elif typ == Transform2D:
        return cast(TVariant, Transform2D(Vector2(1, 0), Vector2(0, 1), Vector2(0, 0)))
    elif typ == Plane:
        return cast(TVariant, Plane(Vector3(0, 0, 1), 0))
    elif typ == Quat:
        return cast(TVariant, Quat(0, 0, 0, 1))
    elif typ == AABB:
        return cast(TVariant, AABB(Vector3(0, 0, 0), Vector3(0, 0, 0)))
    elif typ == Basis:
        return cast(TVariant, Basis(default(Vector3), default(Vector3), default(Vector3)))
    elif typ == Transform:
        return cast(TVariant, Transform(default(Basis), default(Vector3)))
    elif typ == NodePath:
        return cast(TVariant, NodePath("."))
    elif typ == RID:
        return cast(TVariant, RID(0))
    elif typ == Object:
        return cast(TVariant, Object("Object", {}))
    elif typ == Resource:
        return cast(TVariant, Resource("Resource", "res://"))
    elif get_origin(typ) == Array:
        return cast(TVariant, Array(get_args(typ)[0]))
    elif get_origin(typ) == List or typ == list:
        return cast(TVariant, [])
    elif get_origin(typ) == Dict or typ == dict:
        return cast(TVariant, {})
    else:
        return cast(TVariant, None)


def get_value(typ: Type[TVariant], name: str, default_value: Optional[TVariant] = None) -> TVariant:
    return default(typ)
```

**VS Code extension or command-line**
I've tested this code using command-line tool `1.1.134`

**Comment by erictraut** [comment]
Thanks for the issue report.

Your use of constrained type variables with a large number of constraints is difficult for a type checker to handle. I tried it on mypy (after removing the recursive type definition, which isn't supported by mypy), and it simply crashed.

I investigated the performance issue in pylance. We were hitting an n^2 algorithm within the core type checker. I was able to eliminate the n^2 behavior, which improved performance significantly. The analysis time for your sample dropped from 14520ms to 80ms on my machine!

This change will be included in the next release of pylance.
